homeassistant/components/bluetooth/models.py

```diff
@@ -134,6 +151,142 @@ async def async_diagnostics(self) -> dict[str, Any]:
+        if connectable:
+            self._details["connector"] = connector
+            self._expire_seconds = (
+                CONNECTABLE_FALLBACK_MAXIMUM_STALE_ADVERTISEMENT_SECONDS
+            )
+
+    @hass_callback
+    def async_setup(self) -> CALLBACK_TYPE:
+        """Set up the scanner."""
+        return async_track_time_interval(
+            self.hass, self._async_expire_devices, timedelta(seconds=30)
+        )
+
+    def _async_expire_devices(self, _datetime: datetime.datetime) -> None:
+        """Expire old devices."""
+        now = MONOTONIC_TIME()
+        expired = [
+            address
+            for address, timestamp in self._discovered_device_timestamps.items()
+            if now - timestamp > self._expire_seconds
+        ]
+        for address in expired:
+            del self._discovered_device_advertisement_datas[address]
+            del self._discovered_device_timestamps[address]
+
+    @property
+    def discovered_devices(self) -> list[BLEDevice]:
+        """Return a list of discovered devices."""
+        return [
+            device_advertisement_data[0]
+            for device_advertisement_data in self._discovered_device_advertisement_datas.values()
+        ]
+
+    @property
+    def discovered_devices_and_advertisement_data(
+        self,
+    ) -> dict[str, tuple[BLEDevice, AdvertisementData]]:
+        """Return a list of discovered devices and advertisement data."""
+        return self._discovered_device_advertisement_datas
+
+    @hass_callback
+    def _async_on_advertisement(
+        self,
+        address: str,
+        rssi: int,
+        local_name: str | None,
+        service_uuids: list[str],
+        service_data: dict[str, bytes],
+        manufacturer_data: dict[int, bytes],
+        tx_power: int | None,
+    ) -> None:
+        """Call the registered callback."""
+        now = MONOTONIC_TIME()
+        if prev_discovery := self._discovered_device_advertisement_datas.get(address):
+            # Merge the new data with the old data
+            # to function the same as BlueZ which
+            # merges the dicts on PropertiesChanged
+            prev_device = prev_discovery[0]
+            prev_advertisement = prev_discovery[1]
+            if (
+                local_name
+                and prev_device.name
+                and len(prev_device.name) > len(local_name)
+            ):
+                local_name = prev_device.name
+            if prev_advertisement.service_uuids:
+                service_uuids = list(
+                    set(service_uuids + prev_advertisement.service_uuids)
+                )
+            if prev_advertisement.service_data:
+                service_data = {**prev_advertisement.service_data, **service_data}
+            if prev_advertisement.manufacturer_data:
+                manufacturer_data = {
+                    **prev_advertisement.manufacturer_data,
+                    **manufacturer_data,
+                }
```

Comment:
When I added the tests, I realized we weren't merging the service_data/service_uuids/manufacturer_data like BlueZ which means some devices probably were not being discovered properly with esphome.  I haven't actually found any so I didn't mark it as a bugfix, but it could be.